cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs

```diff
@@ -223,26 +226,46 @@ instantiateProjectConfigSkeletonWithCompiler os arch impl _flags skel = go $ map
       (Lit False) -> maybe ([]) ((: []) . go) mf
       _ -> error $ "unable to process condition: " ++ show cnd -- TODO it would be nice if there were a pretty printer
 
-projectSkeletonImports :: ProjectConfigSkeleton -> [ProjectConfigImport]
+projectSkeletonImports :: ProjectConfigSkeleton -> [ProjectConfigPath]
 projectSkeletonImports = view traverseCondTreeC
 
-parseProjectSkeleton :: FilePath -> HttpTransport -> Verbosity -> [ProjectConfigImport] -> FilePath -> BS.ByteString -> IO (ParseResult ProjectConfigSkeleton)
-parseProjectSkeleton cacheDir httpTransport verbosity seenImports source bs = (sanityWalkPCS False =<<) <$> liftPR (go []) (ParseUtils.readFields bs)
+-- | Parses a project from its root config file, typically cabal.project.
+parseProject :: FilePath -> FilePath -> HttpTransport -> Verbosity -> ProjectConfigToParse -> IO (ParseResult ProjectConfigSkeleton)
+parseProject rootConfig = parseProjectSkeleton [] "" (ProjectConfigPath $ rootConfig :| [])
+
+-- | Parses project configuration recursively, following imports.
+parseProjectSkeleton :: [ProjectConfigPath] -> FilePath -> ProjectConfigPath -> FilePath -> HttpTransport -> Verbosity -> ProjectConfigToParse -> IO (ParseResult ProjectConfigSkeleton)
+parseProjectSkeleton [] _ (ProjectConfigPath (rootPath :| [])) cacheDir httpTransport verbosity configToParse =
+  let (projectDir, projectFileName) = splitFileName rootPath; projectPath = ProjectConfigPath $ projectFileName :| []
+   in parseProjectSkeleton [projectPath] projectDir projectPath cacheDir httpTransport verbosity configToParse
+parseProjectSkeleton seenImports dir rootOrImport cacheDir httpTransport verbosity (ProjectConfigToParse bs) =
+  (sanityWalkPCS False =<<) <$> liftPR (go rootOrImport []) (ParseUtils.readFields bs)
   where
-    go :: [ParseUtils.Field] -> [ParseUtils.Field] -> IO (ParseResult ProjectConfigSkeleton)
-    go acc (x : xs) = case x of
-      (ParseUtils.F l "import" importLoc) ->
-        if importLoc `elem` seenImports
-          then pure . parseFail $ ParseUtils.FromString ("cyclical import of " ++ importLoc) (Just l)
+    go :: ProjectConfigPath -> [ParseUtils.Field] -> [ParseUtils.Field] -> IO (ParseResult ProjectConfigSkeleton)
+    go configPath acc (x : xs) = case x of
+      (ParseUtils.F l "import" importLoc) -> do
+        let importLocPath = ProjectConfigPath (importLoc <| coerce configPath)
+        let fullLocPath = fullConfigPathRoot dir importLocPath
+        normLocPath <- canonicalizeConfigPath dir importLocPath
+        normSeenImports <- nub <$> mapM (canonicalizeConfigPath dir) seenImports
+        info verbosity $ "import location, relative: " ++ show importLoc
+        info verbosity $ "import path, relative:\n" ++ showProjectConfigPath importLocPath
+        info verbosity $ "import path, normalized:\n" ++ showProjectConfigPath normLocPath
```

Comment:
The logging is there to help developers understand what is going on. Here's a snippet (using an updated `showProjectConfigPath` that doesn't use leading spaces - these were trimmed by logging unfortunately so I had to avoid their use);

```
import location, relative: ../hops-4.config

import path, relative
=====================
* hops-0.project
-* hops/hops-1.config
--* ../hops-2.config
---* hops/hops-3.config
----* ../hops-4.config

import path, normalized
=======================
* hops-0.project
-* hops/hops-1.config
--* hops-2.config
---* hops/hops-3.config
----* hops-4.config
```

The command I used to log this is;

```
$ pwd
.../cabal/cabal-testsuite/PackageTests/ConditionalAndImport
$ cabal build all --project-file=hops-0.project --verbose=2 2>&1 | tee log.txt

Reply:
fwiw, I also find three forms excessive. It's handy for experimenting with formats, probably, but I'd expect, in the final version of the patch, to retain only one.